constrained_linear_regression/__init__.py

```diff
@@ -78,6 +84,195 @@ def fit(self, X, y, min_coef=None, max_coef=None):
+
+    tol : float, default=1e-15
+        The tolerance for the optimization.
+
+    learning_rate : float, default=1.0
+        The learning rate for the optimization.
+
+    max_iter : int, default=10000
+        The maximum number of iterations for the optimization algorithm.
+
+    Methods
+    --------
+    fit(X, y, horizon, min_coef=None, max_coef=None, initial_beta=None):
+        Fits the model according to the given training data.
+    """
+    global_horizon_count = 0
+    
+    def __init__(
+            self,
+            fit_intercept=True,
+            normalize=False,
+            copy_X=True,
+            nonnegative=False,
+            ridge=0,
+            lasso=0,
+            tol=1e-15,
+            learning_rate=1.0,
+            max_iter=10000,
+            penalty_rate=0
+    ):
+        super().__init__(fit_intercept, normalize, copy_X, nonnegative, ridge, lasso, tol, learning_rate, max_iter)
+        self.penalty_rate = penalty_rate
+
+    def fit(self, X, y, horizon, min_coef=None, max_coef=None, initial_beta=None):
+        X, y = check_X_y(X, y, accept_sparse=['csr', 'csc', 'coo'], y_numeric=True, multi_output=False)
+        X, y, X_offset, y_offset, X_scale = _preprocess_data(
+            X,
+            y,
+            fit_intercept=self.fit_intercept,
+            normalize=self.normalize,
+            copy=self.copy_X,
+        )
+        self.min_coef_ = min_coef if min_coef is not None else np.ones((horizon, X.shape[1]))* -np.inf
+        self.max_coef_ = max_coef if max_coef is not None else np.ones((horizon, X.shape[1]))* np.inf
+        if self.nonnegative:
+            self.min_coef_ = np.clip(self.min_coef_, 0, None)
+
+        if initial_beta is not None:
+            # providing initial_beta may be useful,
+            # if initial solution does not respect the constraints.
+            beta = initial_beta
+        else:
+            beta = np.zeros(X.shape[1]).astype(float)
+
+        prev_beta = beta + 1
+        hessian = np.dot(X.transpose(), X)
+        if self.ridge:
+            hessian += np.eye(X.shape[1]) * self.ridge
+
+        loss_scale = len(y)
+        step = 0
+        while not (np.abs(prev_beta - beta) < self.tol).all():
+            if step > self.max_iter:
+                print('THE MODEL DID NOT CONVERGE')
+                break
+            step += 1
+            prev_beta = beta.copy()
+            for i in range(len(beta)):
+                if self.penalty_rate:
+                    beta[i] = self.update_beta_penalty(beta, i, X, y, hessian, loss_scale, step/self.max_iter)
+                else:
+                    beta[i] = self.update_beta_clip(beta, i, X, y, hessian, loss_scale)
+
+        self.coef_ = beta
+        self._set_intercept(X_offset, y_offset, X_scale)
+        
+        # Goes to the next multi-level model
+        MultiConstrainedLinearRegression.global_horizon_count += 1
```

Comment:
Why do you change a variable on the global class level?
Wouldn't it be a problem if we create two different instances of `MultiConstrainedLinearRegression`? In this case, training the second instance could change the behaviour of the first instance, and this smells very badly.

Reply:
I see, it can go messy, but there's no other way to work with Darts. :( 
That's why I introduced `def reset(self)`